src/fairseq2/recipes/hf/datasets.py

```diff
@@ -0,0 +1,188 @@
+        return self.name
+
+    def create_reader(
+        self,
+        dataset: Dataset,
+        gang: Gang,
+        bucket_size: Optional[int] = None,
+        max_num_elements: Optional[int] = None,
+        max_seq_len: Optional[int] = None,
+        min_seq_len: Optional[int] = None,
+        seq_len_col: Optional[str] = None,
+        num_accumulate: int = 1,
+        num_prefetch: int = 1,
+        seed: int = 2,
+        **extra: Any,
+    ) -> DataPipelineReader[HFBatch]:
+        """
+        Convert HF Dataset into a fairseq2 PipelineReader.
+        The HF-specific processing logic is assumed to be
+        done before the conversion.
+
+        **Batching**: The batching can be done externally or using
+        bucketing strategies in fairseq2. If both ``bucket_size``
+        and ``max_num_elements`` are None, then we assume the
+        ``dataset`` is already batched and do nothing. Otherwise:
+
+        - If ``bucket_size`` is not None, generate batches of fixed size
+        - If ``max_num_elements`` is not None, generate batches up to
+        `max_num_elements` size. In this case, ``max_seq_len`` must be
+        given.
+
+        In addition, the data
+        should be prepared to have a column that contains
+        sequence length for each example.
+
+        This function should not be used for big datatset.
+
+        :param gang:
+            The gang over which to shard the dataset.
+        :param max_num_elements:
+            The maximum number of elements in each batch.
+        :param max_seq_len:
+            The maximum sequence length of each example.
+            Examples longer than this value will be cropped.
+        :param min_seq_len:
+            The minimum sequence length of each example. Examples
+            shorter than this value will be dropped.
+        :param seq_len_col:
+            The column that stores length of source sequence. Must
+            be provided if the data is to be batched by length
+        :param num_accumulate:
+            The number of batches to accumulate in each iteration.
+            Typically used with gradient accumulation during
+            training.
+        :param num_prefetch:
+            The number of batches to prefetch in background.
+        :param seed:
+            The seed to initialize the random number generators
+            used internally.
+        :param extras:
+            The extra parameters specific to the dataset
+            implementation.
+        """
+
+        # Resolve the batching strategies
+        def resolve_batching_strategy() -> int:
+            if bucket_size is None and max_num_elements is None:
+                return 0  # NO_BATCH
+            elif bucket_size is not None and max_num_elements is not None:
+                raise ValueError("Can only batch by size or by length, not both")
+            elif bucket_size is not None:
+                return 1  # BATCH_BY_SIZE
+            else:
+                assert (
```

Comment:
Please use a `ValueError` here similar to line 94. We want to avoid "bad research library habits". `assert` statements should only be used to check some implementation invariants. That means; a failed assert should mean a bug in the implementation. Any error in user-specific inputs should be handled via exceptions.